File: src/main/java/june_2025/BrowserHistory.java
package june_2025;

public class BrowserHistory {

    public static void main(String[] args) {
        BrowserHistory browserHistory = new BrowserHistory("leetcode.com");
        browserHistory.visit("google.com");       // You are in "leetcode.com". Visit "google.com"
        browserHistory.visit("facebook.com");     // You are in "google.com". Visit "facebook.com"
        browserHistory.visit("youtube.com");      // You are in "facebook.com". Visit "youtube.com"
        System.out.println(browserHistory.back(1));                   // You are in "youtube.com", move back to "facebook.com" return "facebook.com"
        System.out.println(browserHistory.back(1));                   // You are in "facebook.com", move back to "google.com" return "google.com"
        System.out.println(browserHistory.forward(1));                // You are in "google.com", move forward to "facebook.com" return "facebook.com"
        browserHistory.visit("linkedin.com");     // You are in "facebook.com". Visit "linkedin.com"
        browserHistory.forward(2);                // You are in "linkedin.com", you cannot move forward any steps.
        System.out.println(browserHistory.back(2));                   // You are in "linkedin.com", move back two steps to "facebook.com" then to "google.com". return "google.com"
        System.out.println(browserHistory.back(7));                   // You are in "google.com", you can move back only one step to "leetcode.com". return "leetcode.com"
    }


    ListNode linkedList;
    ListNode curr;

    public BrowserHistory(String homepage) {
        linkedList = new ListNode(homepage);
        curr = linkedList;
    }

    public void visit(String url) {
        ListNode newNode = new ListNode(url);
        curr.next = newNode;
        newNode.prev = curr;

        curr = curr.next;
    }

    public String back(int steps) {
        while (curr.prev != null && steps > 0) {
            curr = curr.prev;
            steps--;
        }
        return curr.value;

    }

    public String forward(int steps) {
        while (curr.next != null && steps > 0) {
            curr = curr.next;
            steps--;
        }
        return curr.value;
    }

    static class ListNode {
        String value;
        ListNode next;
        ListNode prev;

        public ListNode(String value) {
            this.value = value;
        }
    }

}

File: python/second/browser_history.py
class BrowserHistory(object):

    def __init__(self, homepage):
        """
        :type homepage: str
        """
        self.linkedlist = ListNode(homepage)
        self.curr = self.linkedlist
        

    def visit(self, url):
        """
        :type url: str
        :rtype: None
        """
        newNode = ListNode(url)

        self.curr.next = newNode
        newNode.prev = self.curr

        self.curr = self.curr.next
        

    def back(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        while self.curr.prev != None and steps > 0:
            self.curr = self.curr.prev
            steps -= 1
        return self.curr.val
        

    def forward(self, steps):
        """
        :type steps: int
        :rtype: str
        """
        while self.curr.next != None and steps > 0:
            self.curr = self.curr.next
            steps -= 1
        return self.curr.val
        

class ListNode(object):
    def __init__(self, val, next = None, prev = None):
        self.val = val
        self.next = next
        self.prev = prev